File: ports/php/lipi_core.php
<?php
// lipi — Bengali (বাংলা) DOCX & PDF generation.  https://github.com/bemoshiur/Bengali-Docgen-docx-PDF-production-grade
// Author: S M Moshiur Rahman  <bemoshiur@gmail.com>  ·  +8801717714676 (WhatsApp only)
// Free & open source under the MIT License. Keep this attribution if you use this code.
// lipi core primitives — PHP port.
// odttf embedded-font obfuscation (ECMA-376 §17.8.1) + Bengali numerals and
// South-Asian Taka/Rupee formatting. Full OOXML writer stays TS-first (repo root).
// Run: php lipi_core.php
namespace LipiCore;

const BENGALI_DIGITS = ['০', '১', '২', '৩', '৪', '৫', '৬', '৭', '৮', '৯'];
const TAKA_SIGN = '৳';
const RUPEE_SIGN = '₹';
const SFNT_MAGICS = [0x00010000, 0x4F54544F, 0x74727565, 0x74746366];

function make_font_key(): string
{
    $hex = strtoupper(bin2hex(random_bytes(16)));
    return sprintf(
        '{%s-%s-%s-%s-%s}',
        substr($hex, 0, 8),
        substr($hex, 8, 4),
        substr($hex, 12, 4),
        substr($hex, 16, 4),
        substr($hex, 20, 12)
    );
}

/** 16-byte XOR key: strip {}-, parse 32 hex → 16 bytes, then REVERSE. */
function font_key_to_xor_key(string $fontKey): array
{
    $hex = preg_replace('/[{}\-]/', '', $fontKey);
    if (!preg_match('/^[0-9a-fA-F]{32}$/', $hex)) {
        throw new \InvalidArgumentException("invalid font key: $fontKey");
    }
    return array_reverse(array_values(unpack('C*', hex2bin($hex))));
}

/** XOR the first 32 bytes. Symmetric (obfuscate == deobfuscate). */
function obfuscate(string $ttf, string $fontKey): string
{
    $key = font_key_to_xor_key($fontKey);
    $n = min(32, strlen($ttf));
    $head = '';
    for ($i = 0; $i < $n; $i++) {
        $head .= chr(ord($ttf[$i]) ^ $key[$i % 16]);
    }
    return $head . substr($ttf, $n);
}

function deobfuscate(string $odttf, string $fontKey): string
{
    return obfuscate($odttf, $fontKey);
}

function has_sfnt_magic(string $buf): bool
{
    if (strlen($buf) < 4) {
        return false;
    }
    return in_array(unpack('N', substr($buf, 0, 4))[1], SFNT_MAGICS, true);
}

function to_bengali_numerals($value): string
{
    $out = '';
    foreach (str_split((string) $value) as $c) {
        $out .= ctype_digit($c) ? BENGALI_DIGITS[(int) $c] : $c;
    }
    return $out;
}

/** Last three digits, then groups of two: '1000000' → '10,00,000'. */
function group_south_asian(string $intDigits): string
{
    $s = preg_replace('/^0+(?=\d)/', '', $intDigits);
    if (strlen($s) <= 3) {
        return $s;
    }
    $head = substr($s, 0, -3);
    $tail = substr($s, -3);
    $parts = [];
    while (strlen($head) > 2) {
        array_unshift($parts, substr($head, -2));
        $head = substr($head, 0, -2);
    }
    array_unshift($parts, $head);
    return implode(',', $parts) . ',' . $tail;
}

function format_money($amount, string $symbolChar, string $numerals = 'bengali', bool $symbol = true, ?int $decimals = null): string
{
    $neg = $amount < 0;
    $a = abs($amount);
    if ($decimals !== null) {
        $fixed = number_format($a, $decimals, '.', '');
    } elseif ($a == (int) $a) {
        $fixed = (string) (int) $a;
    } else {
        $fixed = number_format($a, 2, '.', '');
    }
    if (strpos($fixed, '.') !== false) {
        [$ip, $fp] = explode('.', $fixed);
        $body = group_south_asian($ip) . '.' . $fp;
    } else {
        $body = group_south_asian($fixed);
    }
    if ($numerals === 'bengali') {
        $body = to_bengali_numerals($body);
    }
    return ($neg ? '-' : '') . ($symbol ? $symbolChar : '') . $body;
}

function format_taka($amount, string $numerals = 'bengali', bool $symbol = true, ?int $decimals = null): string
{
    return format_money($amount, TAKA_SIGN, $numerals, $symbol, $decimals);
}

function format_rupee($amount, string $numerals = 'bengali', bool $symbol = true, ?int $decimals = null): string
{
    return format_money($amount, RUPEE_SIGN, $numerals, $symbol, $decimals);
}

// ── self-test ───────────────────────────────────────────────────────────────
if (PHP_SAPI === 'cli' && isset($argv[0]) && realpath($argv[0]) === realpath(__FILE__)) {
    $ok = true;
    $check = function (string $name, bool $cond) use (&$ok) {
        echo ($cond ? '  ✓ ' : '  ✗ FAIL: ') . $name . "\n";
        $ok = $ok && $cond;
    };
    $ttf = file_get_contents(__DIR__ . '/../../packages/fonts/fonts/NotoSansBengali-Regular.ttf');
    $key = make_font_key();
    $odttf = obfuscate($ttf, $key);
    $check('input has sfnt magic', has_sfnt_magic($ttf));
    $check('obfuscated loses magic', !has_sfnt_magic($odttf));
    $check('roundtrip is byte-identical', deobfuscate($odttf, $key) === $ttf);
    $check('to_bengali_numerals(1234567)', to_bengali_numerals(1234567) === '১২৩৪৫৬৭');
    $check('format_taka(1000000)', format_taka(1000000) === '৳১০,০০,০০০');
    $check('format_rupee(1000000)', format_rupee(1000000) === '₹১০,০০,০০০');
    echo ($ok ? "PHP PORT OK ✓\n" : "PHP PORT FAILED ✗\n");
    exit($ok ? 0 : 1);
}

File: ports/go/lipicore.go
// lipi — Bengali (বাংলা) DOCX & PDF generation.  https://github.com/bemoshiur/Bengali-Docgen-docx-PDF-production-grade
// Author: S M Moshiur Rahman  <bemoshiur@gmail.com>  ·  +8801717714676 (WhatsApp only)
// Free & open source under the MIT License. Keep this attribution if you use this code.
// lipi core primitives — Go port.
// odttf embedded-font obfuscation (ECMA-376 §17.8.1) + Bengali numerals and
// South-Asian Taka/Rupee formatting. Full OOXML writer stays TS-first (repo root).
// Run: go run lipicore.go   (from the repo root)
package main

import (
	"crypto/rand"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
)

var bengaliDigits = []string{"০", "১", "২", "৩", "৪", "৫", "৬", "৭", "৮", "৯"}

const takaSign = "৳"
const rupeeSign = "₹"

var sfntMagics = map[uint32]bool{0x00010000: true, 0x4F54544F: true, 0x74727565: true, 0x74746366: true}

var reNonKey = regexp.MustCompile(`[{}\-]`)
var reHex32 = regexp.MustCompile(`^[0-9a-fA-F]{32}$`)

func makeFontKey() string {
	b := make([]byte, 16)
	_, _ = rand.Read(b)
	h := strings.ToUpper(fmt.Sprintf("%x", b))
	return fmt.Sprintf("{%s-%s-%s-%s-%s}", h[0:8], h[8:12], h[12:16], h[16:20], h[20:32])
}

// fontKeyToXorKey: strip {}-, parse 32 hex -> 16 bytes, then REVERSE.
func fontKeyToXorKey(fontKey string) ([]byte, error) {
	h := reNonKey.ReplaceAllString(fontKey, "")
	if !reHex32.MatchString(h) {
		return nil, fmt.Errorf("invalid font key: %s", fontKey)
	}
	b := make([]byte, 16)
	for i := 0; i < 16; i++ {
		v, _ := strconv.ParseUint(h[i*2:i*2+2], 16, 8)
		b[i] = byte(v)
	}
	for i, j := 0, 15; i < j; i, j = i+1, j-1 {
		b[i], b[j] = b[j], b[i]
	}
	return b, nil
}

// obfuscate XORs the first 32 bytes. Symmetric (obfuscate == deobfuscate).
func obfuscate(ttf []byte, fontKey string) []byte {
	key, _ := fontKeyToXorKey(fontKey)
	out := make([]byte, len(ttf))
	copy(out, ttf)
	n := 32
	if len(out) < 32 {
		n = len(out)
	}
	for i := 0; i < n; i++ {
		out[i] ^= key[i%16]
	}
	return out
}

func deobfuscate(odttf []byte, fontKey string) []byte { return obfuscate(odttf, fontKey) }

func hasSfntMagic(buf []byte) bool {
	if len(buf) < 4 {
		return false
	}
	magic := uint32(buf[0])<<24 | uint32(buf[1])<<16 | uint32(buf[2])<<8 | uint32(buf[3])
	return sfntMagics[magic]
}

func toBengaliNumerals(v interface{}) string {
	s := fmt.Sprintf("%v", v)
	var b strings.Builder
	for _, c := range s {
		if c >= '0' && c <= '9' {
			b.WriteString(bengaliDigits[c-'0'])
		} else {
			b.WriteRune(c)
		}
	}
	return b.String()
}

// groupSouthAsian: last three digits, then groups of two: "1000000" -> "10,00,000".
func groupSouthAsian(intDigits string) string {
	s := intDigits
	for len(s) > 1 && s[0] == '0' {
		s = s[1:]
	}
	if len(s) <= 3 {
		return s
	}
	head := s[:len(s)-3]
	tail := s[len(s)-3:]
	var parts []string
	for len(head) > 2 {
		parts = append([]string{head[len(head)-2:]}, parts...)
		head = head[:len(head)-2]
	}
	parts = append([]string{head}, parts...)
	return strings.Join(parts, ",") + "," + tail
}

func formatMoney(amount float64, symbolChar, numerals string, symbol bool) string {
	neg := amount < 0
	a := amount
	if neg {
		a = -a
	}
	var fixed string
	if a == float64(int64(a)) {
		fixed = strconv.FormatInt(int64(a), 10)
	} else {
		fixed = strconv.FormatFloat(a, 'f', 2, 64)
	}
	var body string
	if strings.Contains(fixed, ".") {
		p := strings.SplitN(fixed, ".", 2)
		body = groupSouthAsian(p[0]) + "." + p[1]
	} else {
		body = groupSouthAsian(fixed)
	}
	if numerals == "bengali" {
		body = toBengaliNumerals(body)
	}
	sign := ""
	if neg {
		sign = "-"
	}
	sym := ""
	if symbol {
		sym = symbolChar
	}
	return sign + sym + body
}

func formatTaka(amount float64) string  { return formatMoney(amount, takaSign, "bengali", true) }
func formatRupee(amount float64) string { return formatMoney(amount, rupeeSign, "bengali", true) }

func findFont() string {
	dir, _ := os.Getwd()
	for {
		p := filepath.Join(dir, "packages", "fonts", "fonts", "NotoSansBengali-Regular.ttf")
		if _, err := os.Stat(p); err == nil {
			return p
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			break
		}
		dir = parent
	}
	panic("bundled font not found; run from the repo root")
}

func main() {
	ok := true
	check := func(name string, cond bool) {
		if cond {
			fmt.Println("  ✓ " + name)
		} else {
			fmt.Println("  ✗ FAIL: " + name)
			ok = false
		}
	}
	ttf, _ := os.ReadFile(findFont())
	key := makeFontKey()
	odttf := obfuscate(ttf, key)
	check("input has sfnt magic", hasSfntMagic(ttf))
	check("obfuscated loses magic", !hasSfntMagic(odttf))
	check("roundtrip is byte-identical", string(deobfuscate(odttf, key)) == string(ttf))
	check("toBengaliNumerals(1234567)", toBengaliNumerals(1234567) == "১২৩৪৫৬৭")
	check("formatTaka(1000000)", formatTaka(1000000) == "৳১০,০০,০০০")
	check("formatRupee(1000000)", formatRupee(1000000) == "₹১০,০০,০০০")
	if ok {
		fmt.Println("GO PORT OK ✓")
		os.Exit(0)
	}
	fmt.Println("GO PORT FAILED ✗")
	os.Exit(1)
}
